### DISCUSSION TITLE
Welcome to the ATP discussions!

### DISCUSSION THREAD

**Discussion by dholms** [bluesky-social/atproto#100] [Dev Announcements]
We’re using Discussions as a place to hold long-form discussions with our community. We hope that you:

- Play around with the ATP prototype, look through some of the code, and read the architecture overview
- Ask questions about ATP and the technical concepts behind it
- Share your ideas on self-authenticating data structures
- Keep the conversation focused, friendly, and productive. Discussion rules: https://tinyurl.com/bluesky-rules

Comment below if you would like to introduce yourself or go ahead and create a post in another category.

If you prefer synchronous conversation, hop over to our [Matrix room](https://matrix.to/#/#bluesky-dev:matrix.org).

**Comment by dholms** [comment]
Hi all 👋 

I'm Daniel, a protocol engineer & part of the Bluesky team. I care about user controlled data, protocol level sovereignty, and working on an optimistic vision for the web/social networks.

I led development on ADX and am psyched to get this first pass at it out into the world. Take a look, share your thoughts, and stay tuned for what's next 👀 

Thanks for stopping by!

**Comment by pfrazee** [comment]
Hey folks!

I'm Paul, I work in the Bluesky team on the protocol and anything applications-related. I've previously worked on [SSB](https://scuttlebutt.nz/), created the [Beaker Browser](https://github.com/beakerbrowser/beaker), and [live-streamed building a p2p/federated network called CTZN](https://ctzn.network/dev-vlog).

Be sure to upvote my comment so it shows up higher than Daniel's.

**Reply by dholms** [reply]
😒

**Comment by arcalinea** [comment]
Hi I'm Jay, chief cat-herder here. I previously worked on [zksnark cryptocurrencies](https://github.com/arcalinea/bellman-examples), [social apps](happening.net), and [decentralized social apps](https://github.com/arcalinea/bridge-mobile). The one contribution I made to this repo has since been rebased out of existence, but I aspire to write more code again someday. :)

**Comment by austin362667** [comment]
Hi,

I'm Austin, currently working for market making infra.
Really admire your works. it's nice to see such immense progress for bluesky.

Look forward to contribute into the prototype somehow.. Still viewing the codebase.

**Comment by myConsciousness** [comment]
Hi amazing developers,

I'm Kato Shinya, from Japan. I have experience in various front-end and back-end development and recently developed a library wrapping the Twitter API v2.0 in the Dart language, which was published on the official Twitter page.

Well, I have a very strong interest in this project, where do I start?

For starters, I have read through the following documents.
Is it already possible to actually run this in a local environment?

https://atproto.com/guides/overview

Thank you.

**Reply by dholms** [reply]
Hey @myConsciousness 👋 

You can run it locally! 
- clone the repo
- run `yarn` from the root (install dependencies & build)
- run `yarn start` from '/packages/dev-env` to start up a PLC server & PDS locally

From there, you can either sent XRPC requests using curl or programatically. Take a look at the tests in `packages/pds` for some examples of how the client API works.

Let us know if you have any questions :)

**Reply by changeling** [reply]
Hey, @dholms,

I came looking for exactly this information. Could this be added to the toplevel README? I have a slight suspicion I may not be the only one. 😀

**Reply by dholms** [reply]
We need to do a much larger pass on dev tools/docs/READMEs/sdk/etc. And until then we figured we'd hold off on advertising how to get the dev env up and running.

We figured in some ways it's worse to say: "here's how you run dev environment, but good luck after that!" then it was to just not advertise it.

I think I'll hold off on adding it to the landing README for now. It's findable now if you're willing to do some clicking around. Trust that we'll put together some better resources for running the system both as a playground and as a personal server soon :)

**Reply by changeling** [reply]
@dholms I appreciate the reply, and the approach. Are questions welcome at this stage, on a basic environment level, or is it best wait for the docs/tooling/etc to coalesce before tossing them in? And if they're welcome, what might be the best channel for that?

**Reply by dholms** [reply]
@changeling Questions are totally welcome! Especially for folks that are taking the time to dive into the technical docs & code. We just don't want to falsely lure people into thinking this is more ready to run than it is.

Best places are here in GH discussions (or issues if you find something in the code that needs fixing) as well as the [Matrix room](https://matrix.to/#/#bluesky-dev:matrix.org) - although the matrix can sometimes get pretty multithreaded & noisy, so some relative peace for longform discussions can be found here.

**Comment by Nuhvi** [comment]
Hello, I am interested in learning more about how content authoring works technically, and some questions are still hard to answer from reading the docs.

For context, I am coming from a Hypercore background (though not an expert yet) and that biases my thinking into freaking out about multi-device authoring and accidental forks etc.

At the Data repos guides, specifically at the [DID section](https://atproto.com/guides/data-repos#identifier-types) it says:
> ..., but since every user has one data repository

That will probably solve any issues of forking of the linked list of records etc. by just granting a _single_ server the authority to write new records at any given time.

 Is that true? In that case [Federation](https://atproto.com/guides/overview#federation) here is only relevant as mirroring data repos? or just while migrating?
 
 That makes sense too, but I wanted to confirm.
 
 If that is not true, and multiple data repos can write at the same time, then how does conflict resolution is handled?
 
 Am I totally wrong and the user author records at the client side and the data repo is just a server/seeder? I doubt that because that would be just as hard as using Hypercores I guess.
 
 Thanks for your time, and I hope my questions are useful and not totally dumb :)

**Reply by pfrazee** [reply]
Hey @Nazeh, good to see you.

Not a dumb question! We do appoint a specific server as the "home server" and it's responsible for resolving conflicts and producing a linear order. This appointment can be found in the user's DID document. As a consequences, yes, other servers can cache a copy of a repo, but it must be thought of as a cache which could be out of date, since the only node which can canonically assert the current state is the home server.

**Reply by dholms** [reply]
Hey @Nazeh,

Good questions!

Yup content is authored by the server and so it is able to linearize any operations and prevent forks.

The authoritative HEAD of the repo is defined by the server that the user has appointed as its PDS.

The network is federated because there are multiple sources of authority. You are able to choose your PDS or even run your own. There may be some usecases for store & forward between PDSs given the self-authorizing nature of the repos, although to get the freshest copy (current HEAD), you'll always need to refer to the user's PDS.

The scales of power are tilted in the user's favor because the DID registry prevents vendor lockin & allows the user to migrate away from even an adversarial PDS.

**Reply by dholms** [reply]
Oops didn't see @pfrazee's response - same same 😛

**Reply by pfrazee** [reply]
Our answers will be compared for differences by criminal investigators

**Reply by Nuhvi** [reply]
Makes sense, thanks @pfrazee and @dholms 

Another question: it seems from the code that the Trust part of "Self Authentication" is a server-side concern, aiding with caching/mirroring, but also for aggregators/indexers, which means clients entirely trust the PDS/mirror/indexer/aggregator, and don't do any verification to ensure that any data was correctly signed by the corresponding valid `signingKey` at any given point of time, is that correct?

**Reply by pfrazee** [reply]
You could verify in the client if you want, but that's correct that we don't. Your PDS is meant to be your user agent in this network.

**Reply by Nuhvi** [reply]
Finally; access control seems entirely out of scope and the focus is on public data, right?

**Reply by pfrazee** [reply]
It isn't out of scope in the long term, we've had extensive discussions internally about how to do it. Nothing concrete yet.

**Reply by Nuhvi** [reply]
Well if that architecture doesn't scale I guess nothing will. 
DNS dependency and the chosen DID methods leave a lot of sovereignty to be desired, but the rest is a perfectly pragmatic and honest set of choices.

Thank you, and congratulations on the announcement.

**Comment by snarfed** [comment]
Hi all! I'm Ryan, [stereotypical Silicon Valley eng for the last 20+ yrs](https://snarfed.org/resume), last 10 mostly in leadership, currently in climate tech. They mostly stopped letting me write code at work long ago, a bit like @arcalinea, so I do it primarily in spare time, on decentralized social networking projects like this and others in the [IndieWeb](https://indieweb.org/). In those communities, I'm probably best known for [Bridgy](https://brid.gy/). Here, I'm working on implementing ATProto in [Bridgy Fed](https://fed.brid.gy/).
